src/tir/transforms/lower_sparse_tir.cc

```diff
@@ -281,43 +283,120 @@ class IndexTransformer : public StmtExprMutator {
     std::unordered_map<const VarNode*, PrimExpr> var_map;
-    loop_vars.reserve(n_iter);
+    Array<Var> all_loop_vars;
     var_map.reserve(n_iter);
     for (const SpIterVar& sp_iter : sp_block->sp_iter_vars) {
       Var loop_var("v_" + sp_iter->var->name_hint);
-      loop_vars.push_back(loop_var);
+      all_loop_vars.push_back(loop_var);
       var_map[sp_iter->var.get()] = loop_var;
     }
 
-    // Step 3. Create block iters and iter bindings.
-    Array<IterVar> block_iters;
-    Array<PrimExpr> iter_bindings;
-    block_iters.reserve(n_iter);
-    iter_bindings.reserve(n_iter);
-    for (int i = 0; i < n_iter; ++i) {
-      block_iters.push_back(SpIterVarToIterVar(sp_block->sp_iter_vars[i], var_map));
-      iter_bindings.push_back(loop_vars[i]);
-    }
+    // Step 3. Collet block iters and iter bindings.
+    std::set<String> in_stack;
+    in_stack.insert("root");
+    /* A stack that stores block itervars in each block. */
+    std::stack<Array<IterVar>> block_iters_st;
+    /* A stack that stores itervar bindings in each block. */
+    std::stack<Array<PrimExpr>> iter_bindings_st;
+    /* A stack that stores genereated loop vars in each block. */
+    std::stack<Array<Var>> loop_vars_st;
+    /* A stack that stores whether to place init block in each block. */
+    std::stack<bool> place_init_st;
+    /* An indicator that records whether init block has been set. */
+    bool init_set = false;
+    do {
+      /* Block itervars of current block. */
+      Array<IterVar> block_iters;
+      /* Itervar bindings of current block. */
+      Array<PrimExpr> iter_bindings;
+      /* Axis names of current block. */
+      Array<String> axis_names;
+      /* Genereated loop vars of current block. */
+      Array<Var> loop_vars;
+      /* An indicator that records whether there is reduction axis in current block. */
+      bool has_reduction_var = false;
+      for (int i = 0; i < n_iter; ++i) {
+        SpIterVar sp_it_var = sp_block->sp_iter_vars[i];
+        String axis_name = sp_it_var->axis->name;
+        auto&& parent_axis = t_->parent.Get(axis_name);
+        CHECK(parent_axis.defined()) << "Sparse IterVar not defined in Axis Tree.";
+        String parent_axis_name = parent_axis.value();
+        bool is_fixed_axis = sp_it_var->axis->is_fixed();
+        /* Add itervar to current block when
+         * - it's not used yet (not in stack) and
+         *   - it's parent axis was used in outer blocks or
+         *   - it's an iterator to a fixed axis.
+         */
+        if ((is_fixed_axis || in_stack.find(parent_axis_name) != in_stack.end()) &&
+            in_stack.find(axis_name) == in_stack.end()) {
+          loop_vars.push_back(all_loop_vars[i]);
+          axis_names.push_back(std::move(axis_name));
+          block_iters.push_back(SpIterVarToIterVar(sp_it_var, var_map));
+          iter_bindings.push_back(all_loop_vars[i]);
+          has_reduction_var |= sp_it_var->is_reduction;
+        }
```

Comment:
I feels like when a sparse block has three iterators of type `df, dv, df`, the lowered TIR will have loop order of `df, df, dv`?

I think we should break this loop once the `if` isn't satisfied :thinking:. It's not expected to reorder the loops in lowering.

Reply:
Actually I think reorder loops at Sparse TIR stage is not useful. Because all valid reorder schedules can be performed after lowering.